### PR TITLE
coding style: encourage documenting interfaces in the header file

### PR DESCRIPTION
**Pull request by kroening**
The rationale is that users of an interface prefer documentation in the
class or the header file.  They furthermore prefer to hear about the
contract rather than implementation details.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: CODING_STANDARD.md [hunk 1]
```diff
@@ -81,10 +81,15 @@ Formatting is enforced using clang-format. For more information about this, see
...
   necessary.
-- A comment block should immediately precede the definition of the entity it
-  documents, which will generally mean that it will live in the source file.
-  This allows us to take advantage of the one definition rule. If each entity
-  is defined only once, then it is also documented only once.
+- For functions or methods that provide an interface to a module, a comment
+  block should immediately precede the declaration of the entity
+  it documents. This implies that it will live in the header file.
+  The documentions should focus on the externally visible behavior, and
+  should discuss the implementation only to the extend necessary for
```
Comment (peterschrammel) [inline CODING_STANDARD.md]: extent

File: CODING_STANDARD.md [hunk 2]
```diff
@@ -81,10 +81,15 @@ Formatting is enforced using clang-format. For more information about this, see
 - The priority of documentation is readability. Therefore, feel free to use
   Doxygen features, or to add whitespace for multi-paragraph comment blocks if
   necessary.
-- A comment block should immediately precede the definition of the entity it
-  documents, which will generally mean that it will live in the source file.
-  This allows us to take advantage of the one definition rule. If each entity
-  is defined only once, then it is also documented only once.
+- For functions or methods that provide an interface to a module, a comment
+  block should immediately precede the declaration of the entity
+  it documents. This implies that it will live in the header file.
```
Comment (peterschrammel) [inline CODING_STANDARD.md:86]: Note that parameter names cannot be omitted in this case.

File: CODING_STANDARD.md [hunk 3]
```diff
@@ -81,10 +81,15 @@ Formatting is enforced using clang-format. For more information about this, see
...
+  The documentions should focus on the externally visible behavior, and
```
Comment (peterschrammel) [inline CODING_STANDARD.md:87]: Duplicating/splitting documentation doubles the effort of maintaining it and increases the likelihood of outdated documentation.
Does doxygen actually merge these documentations properly?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b5293627d3d9994af059501971720a6afba661f2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89391123

**Review by smowton — APPROVED** [review]
I'd certainly want to see module-user-level docs alongside the interface, then module-developer-level docs in the source file

File: CODING_STANDARD.md (hunk 1 above)
Comment (kroening) [reply CODING_STANDARD.md]: Fixed

File: CODING_STANDARD.md (hunk 3 above)
Comment (kroening) [reply CODING_STANDARD.md:87]: Yes, doxygen does merge these; thus, the documentation in the header should not be repeated.

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b9faa294f09cd1afb1e2e70c98da158d6f609d0c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89432555

File: CODING_STANDARD.md (hunk 3 above)
Comment (allredj) [reply CODING_STANDARD.md:87]: What are `documentions`?

**Pull request merged**